hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -0,0 +1,934 @@
+
+
+@st.composite
+def slices(draw, size):
+    """Generates :xp-ref:`valid slices <indexing.html>` that will select indices
+    up to the supplied ``size``."""
+    # The spec does not specify out of bounds behavior.
+    max_step_size = draw(st.integers(1, max(1, size)))
+    step = draw(
+        st.one_of(
+            st.integers(-max_step_size, -1), st.integers(1, max_step_size), st.none()
+        )
+    )
+    start = draw(st.one_of(st.integers(-size, max(0, size - 1)), st.none()))
+    if step is None or step > 0:
+        stop = draw(st.one_of(st.integers(-size, size)), st.none())
+    else:
+        stop = draw(st.one_of(st.integers(-size - 1, size - 1)), st.none())
+    s = slice(start, stop, step)
+
+    # The spec does not specify behavior for out-of-bounds slices, except for
+    # the case where stop == start.
+    test_list = [0 for _ in range(size)]
+    if (
+        test_list[s] == []
+        and size != 0
+        and start is not None
+        and stop is not None
+        and stop != start
+    ):
+        assume(False)
+
+    return s
+
+
+@defines_strategy()
+def indices(
+    shape: Shape,
+    *,
+    min_dims: int = 0,
+    max_dims: Optional[int] = None,
+    allow_ellipsis: bool = True,
+) -> st.SearchStrategy[BasicIndex]:
+    """Return a strategy for :xp-ref:`valid indices <indexing.html>` of
+    arrays with the specified shape.
+
+    It generates tuples containing some mix of integers, :obj:`python:slice`
+    objects, and ``...`` (an ``Ellipsis``). When a length-one tuple would be
+    generated, this strategy may instead return the element which will index the
+    first axis, e.g. ``5`` instead of ``(5,)``.
+
+    * ``shape`` is the shape of the array that will be indexed, as a tuple of
+      integers >= 0. This must be at least two-dimensional for a tuple to be a
+      valid index;  for one-dimensional arrays use ``xps.slices()`` instead.
+    * ``min_dims`` is the minimum dimensionality of the resulting array from use
+      of the generated index.
+    * ``max_dims`` is the the maximum dimensionality of the resulting array,
+      defaulting to ``min(min_dims + 2, len(shape))``.
+    * ``allow_ellipsis`` specifies whether ``...`` is allowed in the index.
+    """
+    check_type(tuple, shape, "shape")
+    check_argument(
+        len(shape) != 0,
+        "No valid indices for zero-dimensional arrays",
+    )
+    check_argument(
+        all(isinstance(x, int) and x >= 0 for x in shape),
+        f"shape={shape!r}, but all dimensions must be of integer size >= 0",
+    )
+    check_type(bool, allow_ellipsis, "allow_ellipsis")
+    check_type(int, min_dims, "min_dims")
+    check_argument(
+        min_dims <= len(shape),
+        f"min_dims {min_dims} cannot be greater than dimensions of shape {shape!r}",
+    )
+    check_valid_dims(min_dims, "min_dims")
+
+    if max_dims is None:
+        max_dims = min(min_dims + 2, len(shape), NDIM_MAX)
```

Comment:
I had changed this to floor at the shapes dimensions with `min(..., len(shape))` as I dropped newaxis support in accordance with Array API, and so anything greater would not get generated by `BasicIndexStrategy` anyway—maybe performance but really just a clarity fix.

I feel for the Array API extra, it'd be nice actually if `max_dims` just defaulted to `len(shape)`?

Reply:
Yep.  If you read [the `basic_indices()` docs carefully](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.basic_indices), `max_dims` defaults to `max(len(shape), min_dims) + 2` (not `min`!) - so without `allow_newaxis`, it's just `len(shape)`.

Reply:
Hmm, I'm possibly misunderstanding here. Why would we want to generate any `max_dims > len(shape)`, if always `allow_newaxis == False`?

Say we copied how `extra.numpy` does it:
```python
def indices(shape, min_dims=0, max_dims=None):
    ...
    if max_dims is None:
        max_dims = min(max(len(shape), min_dims) + 2, 32)
    print(f"{max_dims=}")
    ...
```
An array of 3 dimensions will get indices that generate for 5 dimensions:
```python
>>> indices((3, 3, 3))
max_dims=5
```
I mean technically it doesn't effect the indices generated because they're always capped at `len(shape)` by way of how `BasicIndexStrategy` works (note you'd need to remove my `check_argument(max_dims <= len(shape), ...)`, but it is unnecessary and potentially confusing behaviour no?